### PR TITLE
Handle malformed origin Uri

### PR DESCRIPTION
**Pull request by jbagga**
Addresses https://github.com/aspnet/Home/issues/2318

cc @Tratcher

### CODE REVIEW

File: src/Microsoft.AspNetCore.Cors/Infrastructure/CorsPolicyExtensions.cs [hunk 1]
```diff
@@ -16,7 +16,9 @@ public static bool IsOriginAnAllowedSubdomain(this CorsPolicy policy, string ori
             {
                 return true;
             }
-            var originUri = new Uri(origin, UriKind.Absolute);
+
+            Uri.TryCreate(origin, UriKind.Absolute, out var originUri);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Cors/Infrastructure/CorsPolicyExtensions.cs]: And if it fails? Does UriHelpers.IsSubdomainOf accept null?

Comment (jbagga) [reply src/Microsoft.AspNetCore.Cors/Infrastructure/CorsPolicyExtensions.cs]: Yeah but it returns false here https://github.com/aspnet/CORS/blob/dev/src/Microsoft.AspNetCore.Cors/Infrastructure/UriHelpers.cs#L12

Comment (Tratcher) [reply src/Microsoft.AspNetCore.Cors/Infrastructure/CorsPolicyExtensions.cs]: https://github.com/aspnet/CORS/blob/c873fc55f9fe8ff5d0de2e54cd763cd4a776c42b/src/Microsoft.AspNetCore.Cors/Infrastructure/UriHelpers.cs#L13 will null ref if you ever get there.

Comment (jbagga) [reply src/Microsoft.AspNetCore.Cors/Infrastructure/CorsPolicyExtensions.cs]: Oh my tests were short circuiting before even going there. Let me look into it. Thanks

**Comment by jbagga** [comment]
@Tratcher Updated

File: src/Microsoft.AspNetCore.Cors/Infrastructure/CorsPolicyExtensions.cs [hunk 2]
```diff
@@ -16,11 +16,20 @@ public static bool IsOriginAnAllowedSubdomain(this CorsPolicy policy, string ori
...
             }
-            var originUri = new Uri(origin, UriKind.Absolute);
-            return policy.Origins
-                .Where(o => o.Contains($"://{_WildcardSubdomain}"))
-                .Select(CreateDomainUri)
-                .Any(domain => UriHelpers.IsSubdomainOf(originUri, domain));
+
+            Uri.TryCreate(origin, UriKind.Absolute, out var originUri);
+
+            if (originUri != null)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Cors/Infrastructure/CorsPolicyExtensions.cs]: Use Uri.TryCreate as the if condition.

File: src/Microsoft.AspNetCore.Cors/Infrastructure/CorsPolicyExtensions.cs [hunk 3]
```diff
@@ -16,11 +16,20 @@ public static bool IsOriginAnAllowedSubdomain(this CorsPolicy policy, string ori
...
+            {
+                return policy.Origins
+                    .Where(o => o.Contains($"://{_WildcardSubdomain}"))
+                    .Select(CreateDomainUri)
+                    .Any(domain => UriHelpers.IsSubdomainOf(originUri, domain));
+            }
+            else
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Cors/Infrastructure/CorsPolicyExtensions.cs]: flatten else, the if returned.

File: test/Microsoft.AspNetCore.Cors.Test/CorsPolicyExtensionsTests.cs [hunk 4]
```diff
@@ -23,6 +23,25 @@ public void IsOriginAnAllowedSubdomain_ReturnsTrueIfPolicyContainsOrigin()
             Assert.True(actual);
         }
 
+        [Theory]
+        [InlineData(null)]
+        [InlineData("null")]
+        [InlineData("http://")]
+        [InlineData("http://*")]
+        [InlineData("http://.domain")]
+        public void IsOriginAnAllowedSubdomain_ReturnsFalseIfOriginIsMalformedUri(string malformedOrigin)
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Cors.Test/CorsPolicyExtensionsTests.cs]: Should we also test some URLs that have something after the authority portion? like: `http://foo.bar/kdkfdkfdkf`

Comment (jbagga) [reply test/Microsoft.AspNetCore.Cors.Test/CorsPolicyExtensionsTests.cs]: This test is specifically for malformed origins. Other tests cover well-formed Uris. I have added a test for the following: `"http://.domain/hello")]`

**Comment by jbagga** [comment]
@rynowak Updated

**Pull request merged**